db/StoreArchive.py

```diff
@@ -0,0 +1,55 @@
+import FetchArchive as fetch
+
+from datetime import datetime
+from pymongo import MongoClient
+import os
+import logging
+
+if not os.path.exists('logs'):
+    os.makedirs('logs')
+logging.basicConfig(filename='logs/StoreArchive.log', level=logging.INFO, format='%(asctime)s:%(levelname)s:%(message)s')
+
+db = MongoClient('localhost', 27017).WallStreetDB
+
+
+def buildIndex():
+    '''Gets, cleans and stores data from archiveIndexGenerator().'''
+    form = '%Y-%m-%dT%H:%M:%SZ'
+    for item in fetch.archiveIndexGenerator():
+        # TODO: Add lastModified field.
+        item['_id'] = item['identifier']
+        key = {'_id': item['_id']}
+        if 'date' in item:  # Clean index date. (str -> datetime)
+            item['datetime'] = datetime.strptime(item['date'], form)
+        # Upsert inserts (instead of update) if the item does not exist.
+        db.ArchiveIndex.update_one(key, {'$set': item}, upsert=True)
+
+
+def buildEpisodes():
+    '''
+    Searches local index for empty episodes (documents with no metadata).
+    Scrapes data and updates document.
+    '''
+    num_failed = 0
+    num_suceed = 0
+    empty_episodes = db.ArchiveIndex.find({'metadata': {'$eq': None}})
+    for i, item in enumerate(empty_episodes):
+        try:
+            episode = fetch.getEpisode(item['_id'])
+            db.ArchiveIndex.update_one({'_id': item['_id']}, {'$set': episode})
+            num_suceed += 1
+        except Exception as e:
+            logging.warning(f'Failed to parse item with id {item["_id"]}')
+            logging.exception(e)
+            num_failed += 1
```

Comment:
ISSUE: Order does not change when the function is re-run, but failed segments appear to be not downloaded.
If episodes with errors are marked by letter, good episodes are marked with '.'
Episodes .A....B.....C...D..E... (Errors are sparse)
After running once:
ABCDE....F....G....H. (Errors are bunched up at the start)

Option 1: Use the .skip(n) method on the cursor (empty_episodes) to skip the first couple episodes with errors.
Option 2: Add a 'failed' field in the document and modify the query to not include the failed episodes. This however will fail if the episode fails due to say network issues and NOT structural issues with the page.

Reply:
With multithreading speed, this is not really a problem anymore, so long as the percentage of expected errors is still small. Currently I observe an error rate of 130/10000  = 1.3%.